### PR TITLE
Added a task manager for the kernel

### PR DESCRIPTION
**Pull request by Zefinder**

### CODE REVIEW

**Review by angeredShrimp — APPROVED** [review]

File: tp_exam/tasks/task_manager.c [hunk 1]
```diff
@@ -0,0 +1,116 @@
+#include "task_manager.h"
+#include "../utils/stacks.h"
+
+struct task_t null_task = {0, 0, 0, 0, 0, 0};
```
Comment (angeredShrimp) [inline tp_exam/tasks/task_manager.c]: Global variable ? No please

Comment (angeredShrimp) [reply tp_exam/tasks/task_manager.c]: Unless no other option

File: tp_exam/tasks/task_manager.c [hunk 2]
```diff
@@ -0,0 +1,116 @@
+#include "task_manager.h"
+#include "../utils/stacks.h"
+
+struct task_t null_task = {0, 0, 0, 0, 0, 0};
+
+struct task_t running_tasks[TASK_NUMBER];
```
Comment (angeredShrimp) [inline tp_exam/tasks/task_manager.c]: Maybe define a fixed address for this structure in memory.h

File: tp_exam/tasks/task_manager.c [hunk 3]
```diff
@@ -0,0 +1,116 @@
+#include "task_manager.h"
+#include "../utils/stacks.h"
+
+struct task_t null_task = {0, 0, 0, 0, 0, 0};
+
+struct task_t running_tasks[TASK_NUMBER];
+int scheduling_task_index;
```
Comment (angeredShrimp) [inline tp_exam/tasks/task_manager.c]: Same, or static

File: tp_exam/tasks/task_manager.c [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+// Inits the task manager
+void init_task_manager() {
+    // We init the available stack index
+    init_task_stack();
+
+    // We init scheduling
+    init_scheduling();
+
+    for (int index = 0; index < TASK_NUMBER; index++) {
```
Comment (angeredShrimp) [inline tp_exam/tasks/task_manager.c]: memset

File: tp_exam/tasks/task_manager.c [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    // TODO Else we ask for a user page
+    // If address is -1 then repush the task index and we return 1
+    // TODO Add in good index PTB
+
+    // We create a task struct with the address within the good index
+    running_tasks[task_index].task_id = next_task_id;
+    running_tasks[task_index].first_page_address = 0x400000;
+    running_tasks[task_index].second_page_address = 0;
+    running_tasks[task_index].has_second_page = 0;
+    running_tasks[task_index].quanta = DEFAULT_QUANTA;
```
Comment (angeredShrimp) [inline tp_exam/tasks/task_manager.c]: quantum

File: tp_exam/tasks/task_manager.c [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    running_tasks[task_index].is_alive = 1;
+
+    // Add 1 to task count
+    next_task_id++;
+
+    return 0;
+}
+
+// Asks for a second page for the specified task id, returns 1 if fails (eg. already has a second page)
+int ask_second_page(uint64_t task_id) {
```
Comment (angeredShrimp) [inline tp_exam/tasks/task_manager.c]: Specify stack page

File: tp_exam/tp.c
Comment (angeredShrimp) [inline tp_exam/tp.c:1]: file removed

File: tp_exam/utils/stacks.c [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+// Pops the task index that is ready to use, removing it from the stack. If none returns -1
+int pop_task_index() {
+    // If no value is pointed (after TASK_NUMBER - 1), we return -1
+    if (task_stack_index == TASK_NUMBER) {
+        return -1;
+    }
+
+    // Else we return the task index value and we increase the task_stack_index
+    int result = task_stack[task_stack_index++];
+    return result;
```
Comment (angeredShrimp) [inline tp_exam/utils/stacks.c]: return task_stack_index == TASK_NUMBER ? -1 : task_stack[task_stack_index++];
do it

**Review by angeredShrimp — COMMENTED** [review]
see comments

File: tp_exam/utils/stacks.c (hunk 7 above)
Comment (Zefinder) [reply tp_exam/utils/stacks.c]: Ew

**Pull request merged**